### PR TITLE
SNOW-1826037: Add a new parameter for temp table cleanup

### PR DESCRIPTION
**Pull request by sfc-gh-jdu**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1826037

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   Enabling the old parameter may break old Snowpark versions.

### CODE REVIEW

File: src/snowflake/snowpark/session.py [hunk 1]
```diff
@@ -577,6 +580,12 @@ def __init__(
             self._conn._get_client_side_session_parameter(
                 _PYTHON_SNOWPARK_AUTO_CLEAN_UP_TEMP_TABLE_ENABLED, False
             )
+            and pkg_resources.parse_version(self.version)
+            >= pkg_resources.parse_version(
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/session.py:584]: @sfc-gh-jdu can you add a unit test to test test with different self.version, and the configrued version number to make sure things works correctly

File: src/snowflake/snowpark/session.py [hunk 2]
```diff
@@ -577,6 +580,12 @@ def __init__(
...
+                self._conn._get_client_side_session_parameter(
+                    _PYTHON_SNOWPARK_AUTO_CLEAN_UP_TEMP_TABLE_ENABLED_VERSION, "0.0.0"
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/session.py:586]: let's make the default empty, when it is empty, means do not enable

**Review by sfc-gh-helmeleegy — APPROVED** [review]

File: src/snowflake/snowpark/session.py (hunk 2 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/session.py]: done

File: src/snowflake/snowpark/session.py (hunk 1 above)
Comment (sfc-gh-jdu) [reply src/snowflake/snowpark/session.py]: added

File: tests/unit/test_session.py [hunk 3]
```diff
@@ -617,3 +618,19 @@ def test_session_builder_app_name_existing_invalid_json_query_tag():
...
+)
+@pytest.mark.parametrize("parameter_name", ["_auto_clean_up_temp_table_enabled"])
+def test_parameter_version(version_value, expected_parameter_value, parameter_name):
+    fake_server_connection = mock.create_autospec(ServerConnection)
+    fake_server_connection._thread_safe_session_enabled = True
+    fake_server_connection._get_client_side_session_parameter.return_value = (
+        version_value
+    )
+    session = Session(fake_server_connection)
+    print(parameter_name)
```
Comment (sfc-gh-aalam) [inline tests/unit/test_session.py:635]:
```suggestion
```

**Review by sfc-gh-yzou — APPROVED** [review]

**Pull request merged**